File: Coding-Patterns/Coding-Patterns-Python/04_dynamic_programming/house_robber.py
# LeetCode 198 - House Robber
# Difficulty: Medium
# Pattern: 1D DP (space-optimized)
#
# Problem: Given an array of non-negative integers representing the
# amount of money in each house, return the maximum amount you can rob
# without robbing two adjacent houses.
#
# Approach: At each house, decide: rob it (prev_prev + current) or skip it (prev).
# Only need to track last two values — no full array needed.
#
# Time: O(n)  Space: O(1)

from typing import List

def rob(nums: List[int]) -> int:
    prev2, prev1 = 0, 0
    for num in nums:
        curr = max(prev1, prev2 + num)
        prev2, prev1 = prev1, curr
    return prev1


# ---- Tests ----
if __name__ == "__main__":
    assert rob([1, 2, 3, 1]) == 4    # rob houses 1 and 3
    assert rob([2, 7, 9, 3, 1]) == 12 # rob houses 1, 3, 5
    assert rob([0]) == 0
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/04_dynamic_programming/HouseRobber.cs
// LeetCode 198 - House Robber
// Difficulty: Medium
// Pattern: 1D DP (space-optimized)
//
// Time: O(n)  Space: O(1)

namespace CodingPatterns.DynamicProgramming;

public class HouseRobber
{
    public int Rob(int[] nums)
    {
        int prev2 = 0, prev1 = 0;
        foreach (var num in nums)
        {
            int curr = Math.Max(prev1, prev2 + num);
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new HouseRobber();

        Console.WriteLine(sol.Rob(new[] { 1, 2, 3, 1 }));       // 4  (1 + 3)
        Console.WriteLine(sol.Rob(new[] { 2, 7, 9, 3, 1 }));    // 12 (2 + 9 + 1)
        Console.WriteLine(sol.Rob(new[] { 5 }));                // 5
        Console.WriteLine(sol.Rob(Array.Empty<int>()));         // 0
    }
}
